src/modals/WalletConnectModal.tsx

```diff
@@ -0,0 +1,194 @@
+/*
+Copyright 2018 - 2022 The Alephium Authors
+This file is part of the alephium project.
+
+The library is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+The library is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with the library. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+import WalletConnectClient, { CLIENT_EVENTS } from '@walletconnect/client'
+import { SessionTypes } from '@walletconnect/types'
+import { useCallback, useEffect, useState } from 'react'
+import styled from 'styled-components'
+
+import Input from '../components/Inputs/Input'
+import { useAddressesContext } from '../contexts/addresses'
+import { useWalletConnectContext } from '../contexts/walletconnect'
+import walletConnectFull from '../images/wallet-connect-full.svg'
+import CenteredModal, { ModalFooterButton, ModalFooterButtons } from './CenteredModal'
+
+interface Props {
+  onClose: () => void
+  onConnect: (client: WalletConnectClient) => void
+}
+
+enum State {
+  Error,
+  InitiateSession,
+  Proposal,
+  RequireUnlock
+}
+
+const WalletConnectModal = ({ onClose, onConnect }: Props) => {
+  const { addresses } = useAddressesContext()
+  const { walletConnect } = useWalletConnectContext()
+  const [uri, setUri] = useState('')
+  const [state, setState] = useState(addresses.length > 0 ? State.InitiateSession : State.RequireUnlock)
+  const [proposal, setProposal] = useState<SessionTypes.Proposal | undefined>(undefined)
+
+  useEffect(() => {
+    walletConnect?.on(CLIENT_EVENTS.session.proposal, async (proposal: SessionTypes.Proposal) => {
+      setProposal(proposal)
+      setState(State.Proposal)
+    })
+
+    walletConnect?.on(CLIENT_EVENTS.session.created, onClose)
+  }, [onClose, walletConnect])
+
+  const onInitiate = useCallback(async () => {
+    walletConnect
+      ?.pair({ uri })
+      .then(() => onConnect(walletConnect))
+      .catch(() => setState(State.Error))
+  }, [walletConnect, uri, onConnect])
+
+  const onApprove = useCallback(async () => {
+    if (proposal === undefined) {
+      setState(State.InitiateSession)
+      return
+    }
+
+    const accounts: string[] = addresses.map((a) => `alephium:${a.network}:${a.hash}`)
+    await walletConnect?.approve({ proposal, response: { state: { accounts } } })
```

Comment:
I should return only an address marked as Main, or the first address if none are otherwise.

Reply:
Maybe add a TODO in the code then, that'd be nice 😇